### PR TITLE
Vega and D3 dependencies and using vega to draw the graph

### PR DESCRIPTION
**Pull request by aghassemi**
Partial for https://github.com/ampproject/amphtml/issues/3991, includes:
- d3 and vega third-party imports
- using vega to draw the graph

Few notes:

- I gave up on npm, managed to get compiling with babel working for npm but closure compiler was a whole different game. Apparently compiling d3 with closure compiler is not an easy feat (https://goo.gl/NTLlIq)
- The third-parties are minified already but they still go through our compile step and probably get a bit more optimized.
- Unfortunately `vega min` is `307k`, `d3 min` is `151k`, so `amp-viz-vega-0.1.js` comes down to `466k` minified. :disappointed:

### CODE REVIEW

**Comment by cramforce** [comment]
WOW (on the code size)

File: extensions/amp-viz-vega/0.1/amp-viz-vega.js [hunk 1]
```diff
@@ -124,15 +136,13 @@ export class AmpVizVega extends AMP.BaseElement {
...
    */
   renderGraph_() {
-    // TODO(aghassemi): Replace with actual rendering implementation.
     return new Promise((resolve, unused) => {
-      const text = 'To be replaced with Vega graph with data: ' +
-          JSON.stringify(this.data_);
-      vsyncFor(this.win).mutate(() => {
-        const textNode = this.element.ownerDocument.createTextNode(text);
-        this.container_.appendChild(textNode);
+      vega.parse.spec(this.data_, (error, chart) => {
```
Comment (cramforce) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: Could we do the JSON parsing? I think that would be good for security.

Comment (aghassemi) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: We are, the `vega.parse` name is misleading. `this.data_` is a parsed JSON object, JSON string comes from either a remote `src` or inline `<script>` and we do the parsing.

That being said, data in Vega spec can also be Urls to other remote Vega spec files and Vega runtime makes the XHR call and does the parsing for those. If that's not something we like, we might be able to manipulate the spec file and inline any Url references by doing the XHR and parsing ourselves.

**Comment by dvoytenko** [comment]
RE:code size. @aghassemi Could you provide any good example of a 3rd party document using Vega?

File: third_party/d3/README.amp [hunk 2]
```diff
@@ -0,0 +1,10 @@
+URL: https://github.com/d3/d3
+License: BSD
+License File: https://github.com/d3/d3/blob/master/LICENSE
+
+Description:
+Copy of minified (d3.min.js) from release v3.5.17
+https://github.com/d3/d3/releases/tag/v3.5.17
+
+Local Modifications:
+None
```
Comment (erwinmombay) [inline third_party/d3/README.amp]: add newline

File: third_party/vega/README.amp [hunk 3]
```diff
@@ -0,0 +1,10 @@
+URL: https://github.com/vega/vega
+License: BSD
+License File: https://github.com/vega/vega/blob/master/LICENSE
+
+Description:
+Copy of minified (vega.min.js) from release v2.6.1
+https://github.com/vega/vega/releases/tag/v2.6.1
+
+Local Modifications:
+None
```
Comment (erwinmombay) [inline third_party/vega/README.amp]: add newline

**Comment by aghassemi** [comment]
@dvoytenko Looks like Wikipedia is one of the main users of Vega. Their <graph> extension is based on Vega ( https://www.mediawiki.org/wiki/Extension:Graph )

File: third_party/d3/README.amp (hunk 2 above)
Comment (aghassemi) [inline third_party/d3/README.amp]: done X 2

File: extensions/amp-viz-vega/0.1/amp-viz-vega.js (hunk 1 above)
Comment (erwinmombay) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: should we be rejecting promise on error?

**Comment by erwinmombay** [comment]
just that last question about the error, off /to @dvoytenko for further review.

very unfortunate with the file size and annoying export issue =/

File: extensions/amp-viz-vega/0.1/amp-viz-vega.js (hunk 1 above)
Comment (aghassemi) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: Good catch. Done

File: extensions/amp-viz-vega/0.1/amp-viz-vega.js [hunk 4]
```diff
@@ -21,6 +21,18 @@ import {isLayoutSizeDefined} from '../../../src/layout';
...
+/*
+ * We are using `require()` instead of `import` here for two reasons:
+ *    1- vega expects d3 to be available on window at import time but babel
+ *       re-orders imports so setting window.d3 happens after loading vega.
+ *    2- d3 and vega are commonJS modules and behaviour of `import *` differs
+ *       between babel and closure compiler (used for dist). In the babel case
+ *       one needs to do x.default but not in the closure compiler case.
+ */
+/* global require: false */
+window.d3 = require('../../../third_party/d3/d3');
```
Comment (dvoytenko) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: @erwinmombay Will this pass through our closure compiler?

Comment (erwinmombay) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: it currently does and it shouldn't be an issue from what I've tested

**Comment by dvoytenko** [comment]
LGTM

File: extensions/amp-viz-vega/0.1/amp-viz-vega.js (hunk 4 above)
Comment (aghassemi) [inline extensions/amp-viz-vega/0.1/amp-viz-vega.js]: I am changing this in the next CL (coming soon)

**Pull request merged**